src/qililab/calibration/calibration_node.py

```diff
@@ -0,0 +1,668 @@
+
+                # Sweep interval:
+                sweep_interval = np.arange(start=0, stop=19, step=1)
+
+                # Extra parameters for this concrete notebook:
+                param1=0
+                param2=0
+                ...
+
+            |
+
+            **2) An experiment/circuit** with its corresponding loops for the sweep given by ``sweep_interval``.
+
+            .. code-block:: python
+
+                # Set the environment and paths:
+                ...
+
+                # Set the platform:
+                platform = ql.build_platform(path=platform_path)
+                platform.connect()
+                platform.initial_setup()
+                platform.turn_on_instruments()
+
+                # Define circuit
+                circuit = ...
+
+                # Loop over the sweeps executing the platform:
+                results_list = []
+                for X in sweep_interval:
+                    platform.set_parameter(alias=alias, parameter=ql.Parameter.X, value=X)
+                    result = platform.execute(program=circuit, num_avg=hw_avg, repetition_duration=repetition_duration)
+                    results_list.append(result.array)
+
+                results = np.hstack(results_list)
+
+            |
+
+            **3) An analysis procedure**, that plots and fits the obtained data to the expected theoretical behaviour and finds the optimal desired parameters.
+
+            .. code-block:: python
+
+                def fit(xdata, results):
+                    ...
+
+                fitted_values, x_data, y_data, figure = fit(xdata=sweep_interval, results=results)
+                plt.show()
+
+            |
+
+            **4) An export data cell**, that calls ``export_calibration_outputs()`` with the dictionary to retrieve from the notebook into the calibration workflow:
+
+            .. code-block:: python
+
+                from qililab.automatic_calibration.calibration_node import export_calibration_outputs
+
+                export_calibration_outputs(
+                    {
+                        "check_parameters": {"x": sweep_interval, "y": results},
+                        "platform_params": [(bus_alias0, qubit, param_name0, fitted_values[0]), (bus_alias1, qubit, param_name1, fitted_values[1])],
+                        "fidelities": [(qubit, "fidelity1", 0.9), (qubit, "fidelity2", 0.95)]  # Fidelities in the output dictionary are optional.
+                    }
+                )
+
+            where the ``check_parameters`` are a dictionary of the saved results to do comparisons against. And the ``platform_params`` are a list of parameters to set on the platform.
+    """
+
+    def __init__(
+        self,
+        nb_path: str,
+        in_spec_threshold: float,
+        bad_data_threshold: float,
+        comparison_model: Callable,
+        drift_timeout: float,
+        qubit_index: int | list[int] | None = None,
+        node_distinguisher: int | str | None = None,
```

Comment:
I would prefer having a `name` argument, such that users can give unique names to all his nodes.

This `name` can also be used inside the `maintain` function of the `CalibrationController` class. This way I feel it is easier to target a single node in the graph.

Reply:
The first part, is literally a discussion we had today with Isaac!

 I agree yes, but I was gonna implement that later on, since we wanted to go to the lab first and we need to finish the comparisons, but ofcourse we don;t need to merge this before that, so we can take it before merging, after the lab when we have more time!
 
I was thinking on being able to pass the `node_id` as argument, and if you do so, it overwrittes the automatic function, and if you don't it takes the name of the notebook as it does now!

Reply:
Respect the second part of the comment, I don;t really think its important, since, the `maintain` is called now to the `Node` instance, which are also individually targetable!

But that would be a super simple change, just changing, that instead than doing the mapping before calling maintain, we do it inside:
![Screenshot 2023-11-08 at 16 58 09](https://github.com/qilimanjaro-tech/qililab/assets/109400222/da6e3d23-a143-4a30-a895-39cebd46a241)

Reply:
If you define the nodes in a loop (like in the example of `CalibrationController`'s docstring) you can only target the nodes of the last iteration.

![image](https://github.com/qilimanjaro-tech/qililab/assets/49232747/e1e7c718-b5e3-424f-a6e0-406d6fb7c4f4)

In the image above, you cannot target the first 2 nodes with the `maintain` function.

Reply:
right, because the example code is wrong! Nice catch!

That will not only not work for maintain, but for anything else!

Reply:
You should instead define first and second as list, where each Node instance for q0 is passed in `first[0]= Node...`

Reply:
Like:
```
import numpy as np
sweep_interval = np.arange(start=0, stop=19, step=1)

# GRAPH CREATION AND NODE MAPPING (key = name in graph, value = node object):
nodes = {}
G = nx.DiGraph()
first, second = [], []

# CREATE NODES:
for qubit in [0, 1]:
    first[qubit] = CalibrationNode(
        nb_path="notebooks/first.ipynb",
        qubit_index=qubit,
        in_spec_threshold=4,
        bad_data_threshold=8,
        comparison_model=norm_root_mean_sqrt_error,
        drift_timeout=1800.0,
    )
    nodes[first[qubit].node_id] = first[qubit]

    second[qubit] = CalibrationNode(
        nb_path="notebooks/second.ipynb",
        qubit_index=qubit,
        in_spec_threshold=2,
        bad_data_threshold=4,
        comparison_model=norm_root_mean_sqrt_error,
        drift_timeout=1.0,
        sweep_interval=sweep_interval,
    )
    nodes[second[qubit].node_id] = second[qubit]

    # GRAPH BUILDING (1 --> 2):
    G.add_edge(first[qubit].node_id, second[qubit].node_id)

# CREATE CALIBRATION CONTROLLER:
controller = CalibrationController(node_sequence=nodes, calibration_graph=G, runcard=path_runcard)

### EXECUTIONS TO DO:
controller.maintain(second[0])
```

Reply:
Changed!

Reply:
Actually, Isaac, made me realize you could access the node as it was before, doing:
- maintain(nodes["first_q0"])
- maintain(controller.node_sequence["first_q0"])